File: dice.cpp
// UCF HSPT 2023 - Dencker's Dice?
// Author - Tyler Marks

// Includes all important libraries
#include <bits/stdc++.h>
// Makes it so we don't have to keep calling "std::"
using namespace std;

// Max Value of the problem we can be asked to solve for
const int MAXVAL = 1'000'000;

// General Solution Idea:
//      For the target t and number of sides d, the probability of reaching that
//      target is p_t = r_1 * p_(t-1) + r_2 * p_(t-2) + ... + r_d * p_(t-min(d, t))
//      where r_i is the probability of rolling i and p_t is the probanility
//      of reaching value t. 

//      Idea 1: Solve for t = 0 to MAXVAL. Since the current t value only depends
//      on t values less than it, we can use dynamic programming. Loop through
//      each possible t value. For the current t value, loop from the min of 0 and t-d
//      up to t-1 and take that p value, multiply by the probability of rolling i and add to p_t.
//      This is however too slow since t can be up to 10^6 and d can be up to 10^6.

//      Since we are working with a fair die, p_1 == p_2 == ... == p_d.
//      We will represent this value as p_roll which equals 1 / d.
//      So p_t = the sum of p_roll * p_(t-i) where i is in the range 1 to min(d, t).
//      We use min(d, t) to make sure (t-i) does not go negative.
//      Since we have the same constant multiplied by a series of terms being added 
//      together, we can factor out this constant.
//      This gives us p_t = p_roll * (p_(t-1) + p_(t-2) + ... + p_(t-min(d, t)))
//      So we have a constant times a series of contiguous terms being added together.

//      Idea 2: We now have a contiguous sum of elements. We can use a prefix sum to
//      spped up our dp. As we loop through t, also expand the prefix sum. Now instead
//      of looping through each possible dice roll, we can take the sum from the prefix
//      sum and use that instead.

int main() {

    // Scan in the number of sides and queries.
    int sides, queries; 
    cin >> sides >> queries;

    // Initialize DP and Prefix Sum arrays
    vector<double> dp(MAXVAL + 1, 0);
    vector<double> prefixSums(MAXVAL+1, 0);

    // The probability of rolling a target of 0 is 100%
    dp[0] = prefixSums[0] = 1;

    // Loop through each target
    for(int cur = 1; cur <= MAXVAL; cur++) {

        // Determine how far back into our prefix sum we must go
        double minState = max(0, cur - sides);

        // Get the sum from 0 to the current target minus 1
        double prevStates = prefixSums[cur-1];

        // If our furthest back is greater than 0, subtract out 0 to minstate - 1
        if(minState > 0) prevStates -= prefixSums[minState - 1];

        // Update the current probability
        dp[cur] = 1.0 / sides * prevStates;

        // Update the prefix sum.
        prefixSums[cur] = prefixSums[cur-1] + dp[cur];
    }

    // Go through each query target and print out its probability
    for(int curQuery = 0; curQuery < queries; curQuery++) {
        int target; cin >> target;
        cout << setprecision(10) << fixed << 100.0 * dp[target] << endl;
    }
}

File: dice.py
# Author : Juan Moscoso

'''

let prob(i, s) be the probability that that we land on a target value i with a fair s-sided die

we can jump to target value i from values i - s to i - 1 all with probability 1 / s because for 
each value from i - s to i - 1, we only have one side of our die that gets us to i and we have s sides
so we have i / s probablity of landing on that side. 

prob(i, s) = (prob(i - s, s) / s) + (prob(i - s + 1, s) / s ) + ... + (prob(i - 1, s) / s)
simplifying we get
prob(i, s) = prob(i - s, s) + prob(i - s + 1, s) + ... + prob(i - 1, s)
	      ---------------------------------------------------------
			 		s
									
However, looping through all from i - s to i - 1 for all i from 1 to our max target value
is too slow. This is O(s * max_value_of_t + v). This is about 10^12 operations which is way too slow.
However, we can speed this up with a prefix sum.

let dp(i, s) be the sum(prob(i, s)) for non-negative values <= i

We can calculate dp(i, s) using the same formula for prob(i, s)

dp(i, s) = dp(i - 1, s) + prob(i, s)
dp(i, s) = dp(i - 1, s) + (prob(i - s, s) + prob(i - s + 1, s) + ... + prob(i - 1, s)) / s

we know that 
prob(i - s, s) + prob(i - s + 1, s) + ... + prob(i - 1, s)) = (dp(i - 1, s) - dp(i - s - 1, s)) / s

dp(i, s) = dp(i - 1, s) + (dp(i - 1, s) - dp(i - s - 1, s)) / s

if we calculate dp(i, s) in increasing i, we can calculate each dp(i, s) for all i in O(1)

Now for every query now, prob(i, s) is equal to dp(i, s) - dp(i - 1, s)
by what we defined dp(i, s) to be earlier

dp(i, s) = prob(0, s) + prob(1, s) + ... + prob(i - 1, s) + prob(i, s)
dp(i - 1, s) = prob(0, s) + prob(1, s) + ... + prob(i - 1, s)

dp(i, s) - dp(i - 1, s) = (prob(0, s) + prob(1, s) + ... + prob(i - 1, s) + prob(i, s)) - (prob(0, s) + prob(1, s) + ... + prob(i - 1, s))
						= prob(0, s) - prob(0, s) + prob(1, s) - prob(1, s) + ... + prob(i - 1, s) - prob(i - 1, s) + prob(i, s)
cancel out like terms   = prob(i, s)

So now, we can solve our queries in O(1), this is assuming we calculate all values of dp(i, s)

'''
s, v = map(int, input().split())
MAX_VALUE = 10**6

# calculate dp for all possible target values
dp = [-1] * (MAX_VALUE + 1)
# the probability that we land on 0 is always 100%
dp[0] = 100
for i in range(1, MAX_VALUE + 1):
	dp[i] = dp[i - 1] + (dp[i - 1] / s)
	# edge case if i - s - 1 < 0 then we shouldnt subtract off dp(i - s - 1, s)
	# because we can reach i from all previous values
	if i - s - 1 >= 0:
		dp[i] -= dp[i - s - 1] / s

for i in range(v):
	target_value = int(input())
	# output prob(i, s) which we know is equal to 
	# dp(i, s) - dp(i - 1, s)
	print(dp[target_value] - dp[target_value - 1])
